### ISSUE TITLE
Valid or same padding?

### ISSUE THREAD

**Issue by brod4910** [facebookarchive/caffe2#2231]
Hello, I have been implementing a convolutional neural network in caffe2 recently and have not found a way to use 'valid' or 'same' padding without changing the source code. Is there a way to specify which type of padding I would like to use when creating a pooling layer or conv layer? I am using python.

**Comment by brod4910** [comment]
After further inspection of the source code, I have come to the conclusion, that 'Valid' padding is setting padding = 0 and 'Same' padding is allowing ConvPoolOpBase to manually calculated the correct padding. I would still appreciate some confirmation on my assumption!

**Comment by yinghai** [comment]
@roadhouse

**Comment by CarlosYeverino** [comment]
Hi @brod4910 ,

did you get an answer for this topic? I also want to use valid or same padding. Do you know how could I set both options?

**Comment by brod4910** [comment]
Hello @CarlosYeverino !

I actually did find my answer, but I ended up having to create my own function that calculated the padding for me whenever I wanted valid or same padding. Currently, I’m not near my computer but I can send you the code tomorrow, if you’d like? If you do not want to wait that long I suggest searching on google “how to calculate same or valid padding” there should either be a formula or code that you’ll be able to use!

**Comment by CarlosYeverino** [comment]
Hello @brod4910 ,

sure, I would like to check out your code solution too! Which caffe2 files did you modify? 

I also made a research and found out that VALID and SAME padding can be set with the parameter legacy_pad in convolution and pooling layers:
```
legacy_pad=1 #VALID
legacy_pad=2 #SAME
```

Example:
`conv1 = brew.conv(model, 'img_data', 'conv1', dim_in=1, dim_out=20, kernel=5, stride=1, legacy_pad=1)` 

Python code run succesfully, however, I haven't tested SAME padding since I don't have a proper case in my mind. I found an issue from last year stating that it wasn't worked well https://github.com/caffe2/caffe2/issues/906
It would be very nice if you can test legacy_pad in your implementations to confirm that it works.

Sources:
https://github.com/pytorch/pytorch/blob/master/caffe2/operators/conv_pool_op_base.h
https://github.com/pytorch/pytorch/blob/master/caffe2/proto/caffe2_legacy.proto
